# Illumina exome sequencing reads mapping scons pipeline

A scons based pipeline for mapping Illumina paired end reads obtained from exome sequencing to the human genome.

### Software Dependencies:
* [scons](http://scons.org/)
* [bwa](http://bio-bwa.sourceforge.net/)
* [samtools](http://www.htslib.org/)
* [Genome Analysis Toolkit](https://software.broadinstitute.org/gatk/)
* [bedtools](http://bedtools.readthedocs.io/)


### Annotation dependencies:

A directory specified by the referenceDir argument containing the following files:

* fasta file of the reference human genome (e.g. reference.fa)

* index of the human reference genome fasta file. It can be generated by the command `samtools faidx reference.fa`

* reference.dict file generated by the command `java -jar CreateSequenceDictionary.jar R=reference.fasta O=reference.dict ` 

* bwa index of the reference genome obtained by the command `bwa index reference.fa`. It produces a bunch of files with the following extensions: .amb .ann .bwt .pac .sa.

An annotation directory specified by the annotationDir argument containing the following file:

* the dbSNP vcf and the vcf index e.g. All_20151104.vcf.idx.

* the .bed file with the exome targeted regions. 




### Running the pipeline:


1. Copy SConstruct, settings.py and sam_remove_soft_clipping.py to the directory where the analysis will be run.

2. Parameters can be set editing the settings.py file or specifiying arguments to scons. In this case parameters from the settings.py file will be ovverriden, see examples below.

2. To start an analysis just launch the scons command from the directory containing the SConstruct file. 


### Examples:

```bash

#Launch the pipeline using all parameters from the settings.py file
scon

#Explicity specify the directory containing the fasta file of the reference genome
scons referenceDir=~/reference

```
