

#Dockerized MuTect 2 SCons wrapper
A very basic SCons wrapper to run [MuTect 2](https://software.broadinstitute.org/gatk/gatkdocs/org_broadinstitute_gatk_tools_walkers_cancer_m2_MuTect2.php) using docker image alexcoppe/gatk.


### Software Dependencies:
* [SCons](http://scons.org/)
* [Docker](https://www.docker.com/)


### Annotation dependencies:

An annotation directory specified by the annotationDir argument containing the following file:

* the dbSNP vcf and the vcf index e.g. All_20151104.vcf.idx.

* cosmic variants vcf and vcf.tbi file (optional)

* cosmic non coding variants vcf and vcf.tbi file (optional)

A directory specified by the bamDir argument containing the following files:


* bam file for normal matched sample

* bam file for tumor matched sample

A directory specified by the genomeDir argument containing:

* the fasta file with the human genome

* index of the human reference genome fasta file. It can be generated by the command `samtools faidx reference.fa`


### Running the pipeline:


1. Copy SConstruct, settings.py to the directory where the analysis will be run.

2. To start an analysis just launch the scons command, with desired parameters, from the directory containing the SConstruct file


### Examples:

```bash

scons data=$(pwd) genome=~/genomic_data/human_genome/ annotation=~/annotation cosmic=CosmicCodingMuts_v77.vcf.gz dbsnp=All_20151104.vcf target=nexterarapidcapture_exome_targetedregions_v1.2.bed nb=n.bam tb=t.bam

```


### List of parameters:

| Argument Name        | Description| Type | Default value
| ------------- |:-------------| :-------------| :--------------|
| bamDir      | The directory containing the bams for tumor-normal matched samples | String | data |
| genomeDir  | Path to the directory containing the reference sequence (e.g. ~/genome ) | String | genome |
| annotationDir | Path to the directory with annotation files (e.g. ~/annotation ) | String | annotation
| cosmic | Name of the vcf file with variants downloaded from Cosmic database. Should be placed in the annotationDir directory | String | cosmic.vcf.gz  |
| dbsnp | Name of the vcf file downloaded from dbSNP database | String | dbsnp.vcf.gz |
| targetRegions | Bed file with exome caputered regions. This file must be located in the annotaion directory | String | target |
| normalBam | Bam file for the normal matched sample | String | normal.bam |
| tumorBam | Bam file for the tumor matched sample | String | tumor.bam |


### Results:

* mutect2-variants.vcf: VCF files with somatic variants annotated with cosmic and dbSNP annotation. 
