# Somatic variants identification with MuTect 2, bam splitted by chromosome

A scons based pipeline for mapping somatic variants identification using MuTect2. Original BAM files is splitted by chromosome to parallelize the process.

<img src="mutect2-split-chr.png" width="900px">

### Software Dependencies:
* [scons](http://scons.org/)
* [SnpEff](http://snpeff.sourceforge.net/)
* [Docker](http:/www.docker.com/)

### Annotation dependencies:

An annotation directory specified by the annotationDir argument containing the following file:

* the dbSNP vcf and the vcf index e.g. All_20151104.vcf.idx.

* [Clinvar](http://www.ncbi.nlm.nih.gov/clinvar/) vcf and clinvar vcf.tbi file

* [Cosmic](http://cancer.sanger.ac.uk/cosmic) coding variants vcf and vcf.tbi file

* [Cosmic](http://cancer.sanger.ac.uk/cosmic) non coding variants vcf and vcf.tbi file

A directory specified by the referenceDir argument containing the following files:

* fasta file of the reference human genome (e.g. reference.fa)

* index of the human reference genome fasta file. It can be generated by the command `samtools faidx reference.fa`

* reference.dict file generated by the command `java -jar CreateSequenceDictionary.jar R=reference.fasta O=reference.dict ` 



### Running the pipeline:


### Examples:


### List of parameters:


| Argument Name        | Description|
| ------------- |:-------------|
| reference      | The fasta file with the human genome|
| referenceDir      | Path to the directory containing the reference|
| annotationDir | Path to to the directory containing annotation files|
| exomeRegions | Exome sequencing kit bed file (located in annotationDir)|



### Results:
