# Somatic variants identification with MuTect 2, bam splitted by chromosome

A scons based pipeline for somatic variants identification in matched tumor-normal pairs using MuTect2. Original .BAM files are splitted by chromosome to parallelize the process.

<img src="mutect2-split-chr.png" width="900px">

### Software Dependencies:
* [scons](http://scons.org/)
* [SnpEff](http://snpeff.sourceforge.net/)
* [Docker](http:/www.docker.com/)

### Annotation dependencies:

An annotation directory specified by the annotationDir argument containing the following file:

* [dbSNP](http://www.ncbi.nlm.nih.gov/SNP/) vcf and the vcf index e.g. All_20151104.vcf.idx.

* [Clinvar](http://www.ncbi.nlm.nih.gov/clinvar/) vcf and clinvar vcf.tbi file

* [Cosmic](http://cancer.sanger.ac.uk/cosmic) coding variants vcf and vcf.tbi file

* [Cosmic](http://cancer.sanger.ac.uk/cosmic) non coding variants vcf and vcf.tbi file

A directory specified by the referenceDir argument containing the following files:

* fasta file of the reference human genome (e.g. reference.fa)

* index of the human reference genome fasta file. It can be generated by the command `samtools faidx reference.fa`

* reference.dict file generated by the command `java -jar CreateSequenceDictionary.jar R=reference.fasta O=reference.dict ` 



### Running the pipeline:

1. Copy SConstruct, settings.py to the directory where the analysis will be run.

2. Parameters can be set editing the settings.py file or specifiying arguments to scons. In this case parameters from the settings.py file will be ovverriden, see examples below.

2. To start an analysis just launch the scons command from the directory containing the SConstruct file. 


### Examples:
```bash

#Launch the pipeline using all parameters from the settings.py file
scon

#Explicitly specify the directory containing the fasta file of the reference genome
scons referenceDir=~/reference

#Analyze only chromosomes 5,6 and 7
scons startChr=5 endChr=7

```


### List of parameters:


| Argument Name        | Description| Type |
| ------------- |:-------------| :-------------|
| reference      | The fasta file with the human genome| String |
| referenceDir      | Path to the directory containing the reference| String |
| annotationDir | Path to to the directory containing annotation files| String |
| exomeRegions | Exome sequencing kit bed file (located in annotationDir)| String |
| dbsnpVCF | The name of the vcf dbSNP vcf file (located in annotationDir)| String |
| tumorBam | BAM obtained from "tumoral" sample| String |
| normalBam | BAM obtained from "normal" sample| String |
| startChr | First chromosome to analyze. Used to identify variants only in a subset of chromosomes | Int |
| endChr | Last chromosome to analyze. Used to identify variants only in a subset of chromosomes | Int |
| snpeffDir | Path to directory containing the SnpEff jar file | String |
| snpeffGenomeVersion |  The version of the genome to be used by snpEff (the one specified with the -v option to snpEff.jar) | String |
|clinvarVCF | Name of the vcf file downloaded from Clinvar database | String |
| cosmicCodingVCF | Name of the vcf file with coding variants downloaded from Cosmic database | String |
| cosmicNonCodingVCF | Name of the vcf file with non coding variants downloaded from Cosmic database | String |



### Results:

* The somatic variants for each chromosome will be in a file called mutect_variants_chrXX_pass_snpEFF_clinvar_cosmic_coding_non_coding.vcf with XX replaced by chromosome name

* merged_variants_sorted.vcf: somatic variants for all analyzed chromosomes
