# Somatic variants identification with MuTect, scons pipeline

A scons based pipeline for stomatic variants identification using [MuTect](https://www.broadinstitute.org/cancer/cga/mutect).

<p align="center">
  <img src="mutect.png" >
</p>

The following steps will be carried out:

1. **MuTect**: somatic variants identification. The following parameters will be used:
    * --analysis_type MuTect
    * --reference_sequence
    * --dbsnp
    * --intervals
    * --input_file:normal
    * --input_file:tumor
    * --vcf
    * --coverage_file
    * --out

2. **Filter out** variants tagged as **REJECT** by MuTect (grep)

3. Variants **effect prediction** and annotation by SnpEff

4. **ClinVar** variants annotation using snpSift

5. **Cosmic** variants annotations (both coding and non-coding)

### Software Dependencies:
* [scons](http://scons.org/)
* [mutect](https://www.broadinstitute.org/cancer/cga/mutect)
* [SnpEff](http://snpeff.sourceforge.net/)
* [SnpSift](http://snpeff.sourceforge.net/SnpSift.html)


### Annotation dependencies:

An annotation directory specified by the annotationDir argument containing the following file:

* the dbSNP vcf and the vcf index e.g. All_20151104.vcf.idx.

* clinvar vcf and clinvar vcf.tbi file

* cosmic coding variants vcf and vcf.tbi file

* cosmic non coding variants vcf and vcf.tbi file

A directory specified by the referenceDir argument containing the following files:

* fasta file of the reference human genome (e.g. reference.fa)

* index of the human reference genome fasta file. It can be generated by the command `samtools faidx reference.fa`

* reference.dict file generated by the command `java -jar CreateSequenceDictionary.jar R=reference.fasta O=reference.dict ` 





### Running the pipeline:


1. Copy SConstruct, settings.py to the directory where the analysis will be run.

2. Parameters can be set editing the settings.py file or specifiying arguments to scons. In this case parameters from the settings.py file will be ovverriden, see examples below.

2. To start an analysis just launch the scons command from the directory containing the SConstruct file. 


### Examples:

```bash

#Launch the pipeline using all parameters from the settings.py file
scon

#Explicitly specify the directory containing the fasta file of the reference genome
scons referenceDir=~/reference

#Explicitly specify 3 parameters
scons referenceDir=~/reference mutectDir=~/local/mutect snpeffDir=~/local/snpEFF

```


### List of parameters:

| Argument Name        | Description| Type |
| ------------- |:-------------| :-------------|
| reference      | The fasta file with the human genome| String |
| referenceDir | Path to the directory containing the reference sequence (e.g. ~/genome ) | String |
| annotationDir | Path to the directory with annotation files (e.g. ~/annotation ) | String |
| exomeRegions | Bed file with exome caputered regions. This file must be located in the annotaion directory | String |
| mutectDir | Path to the directory to containing the muTect-1.1.5.jar file | String |
| dbsnpVCF | Name of the vcf file downloaded from dbSNP database | String |
| clinvarVCF | Name of the vcf file downloaded from Clinvar database | String |
| cosmicCodingVCF | Name of the vcf file with coding variants downloaded from Cosmic database | String |
| cosmicNonCodingVCF | Name of the vcf file with non coding variants downloaded from Cosmic database | String |
| bamN | Bam file from the normal matched sample | String |
| bamT | Bam file from the tumor matched sample | String |
| snpeffDir | Path to the directory containgin the snpEff.jar and SnpSift.jar files | String |
| snpeffGenomeVersion | The version of the genome to be used by snpEff (the one specified with the -v option to snpEff.jar) | String |

SnpEff software should be already installed and configured.


### Results:

* coverage.wig.txt: **Coverage / WIGGLE files** so-called "wiggle" files (in WIGGLE format) that contain useful information about the read coverage observed in the data. This format indicates for every base whether it is sufficiently covered in the tumor and normal to be sensitive enough to call mutations. (description from [gatk forum](http://gatkforums.broadinstitute.org/gatk/discussion/4231/what-is-the-output-of-mutect-and-how-should-i-interpret-it))

* mutect-confident-somatic-variants-with-eff-clinvar-cosmiccod-cosmicnoncod.vcf: the vcf files with somatic variants annotated with clinvar and cosmic. dbSNP annotation comes from the bwa-mem-single-sample pipeline. 

* mutect-variants.txt: the variants text file produced by mutect. See [this page](http://gatkforums.broadinstitute.org/gatk/discussion/4231/what-is-the-output-of-mutect-and-how-should-i-interpret-it) for a description of the columns in the file.

* snpEff_summary.html & snpEff_genes.txt: output from snpeEff, see [this page](http://snpeff.sourceforge.net/SnpEff_manual.html#outputSummary) for a description this two files.

* mutect-variants.vcf: original vcf file obtained from MuTect. Not annotated.
