# Illumina exome sequencing reads mapping scons pipeline

A scons based pipeline for mapping Illumina paired end reads obtained from exome sequencing to the human genome.


### Software Dependencies:
* [scons](http://scons.org/)
* [cutadapt](http://cutadapt.readthedocs.io) (optional)
* [bwa](http://bio-bwa.sourceforge.net/)
* [samtools](http://www.htslib.org/)
* [picard](http://broadinstitute.github.io/picard/)
* [Genome Analysis Toolkit](https://software.broadinstitute.org/gatk/)
* [bedtools](http://bedtools.readthedocs.io/)


### Annotation dependencies:

A directory specified by the *referenceDir* argument containing the following files:

* fasta file of the reference human genome (e.g. *reference.fa*)

* index of the human reference genome fasta file. It can be generated by the command `samtools faidx reference.fa`

* reference.dict file generated by the command `java -jar CreateSequenceDictionary.jar R=reference.fasta O=reference.dict ` 

* bwa index of the reference genome obtained by the command `bwa index reference.fa`. It produces a bunch of files with the following extensions: *.amb .ann .bwt .pac .sa*.

An annotation directory specified by the *annotationDir* argument containing the following file:


* the *.bed* file with the exome targeted regions. 

### Running the pipeline on single sample:


1. Copy *SConstruct*, *settings.py* to the directory where the analysis will be run.

2. Parameters can be set editing the *settings.py* file or specifiying arguments to *scons*. In this case parameters from the *settings.py* file will be ovverriden, see examples below.

3. To start an analysis just launch the *scons* command from the directory containing the *SConstruct* file. 

### Running the pipeline on multiple samples:

1. Every Paired-End reads should be in their own directory and should be named *1.fastq.gz 2.fastq.gz*. Parameters can be set editing the *settings.py* file.

2. To start an analysis just launch launch_bwa_mem.py  the *scons* command from the directory containing the *SConstruct* file and using the *-s* parameter to indicate the directory containing *SConstruct* and *settings.py*.

### Examples:

```bash

#Launch the pipeline using all parameters from the settings.py file
scons

#Explicitly specify the directory containing the fasta file of the reference genome
scons referenceDir=~/reference

#Explicitly specify 3 parameters
scons referenceDir=~/reference gatkJarDir=~/local/GATK sampleName=patient1

#Launch on multiple samples. The current directory should contain a directory for every sample with 1.fastq.gz 2.fastq.gz files in it
~/bwa-mem/launch_bwa_mem.py  -s /home/user/exome-analyses 

```


### List of parameters:

| Argument Name        | Description| Type | Default value |
| ------------- |:-------------| :-------------| :-------------|
| reference      | The fasta file with the human genome| String | reference.fa |
| referenceDir | Path to the directory containing the reference sequence (e.g. ~/genome ) | String | ~/annotation |
| reads1 & reads2 | The the name of the fastq files containing the paired end reads | String | 1.fastq.gz & 2.fastq.gz |
| annotationDir | Path to the directory with annotation files (e.g. ~/annotation ) | String | ~/annotation |
| exomeRegions| The bed file with exome caputered regions. This file must be located in the annotaion directory | String | nexterarapidcapture_exome_targetedregions_v1.2.bed |
| gatkJarDir | The path to the directory to containing the GenomeAnalysisTK.jar file | String | ~/local/GenomeAnalysisTK |
| picardDir | The path to the directory to containing the picard.jar file | String | ~/local |
|dbsnpVCF | dbSnp's vcf file | String | All_20170710.vcf  |
| projectName |  A name for the project | String | m2 |
| sampleName | The sample's name | String | m2 |
| processors | Number of CPUs to be used | Int | 8 |
| maxMemory | Max memory to be used (in bytes) | Int | 1000000000 |
| removeNs | If should trim flanking Ns from reads (default n) | String (y/n) | n |
| tools | Choose between samtools and picard | samtools | samtools |
|samQuality| samtools -q parameter: minimum mapping quality  |Int| 0 |
|requiredFlag| samtools required flag, 0 for unset (for example 2) | Int | 0 |
|filteringFlag| samtools filtering flag, 0 for unset (for example 256) | Int | 0 |

### Results:

The pipeline will produce a bam file called *05_mapping-rmdup-cleaned-realigned-recalibration.bam* and a file called *coverage-hist.txt* with coverage statistics containing the following columns:

1. A first column containing only the word all, it derives from the bedtools command used to produce the file.

2. The coverage level.

3. Number of bases with the coverage level indicated by the previous column.

4. The total number of bases targeted by the exome sequencing.

5. The fraction of bases with coverage equal to the one indicated in column 2

